src/lightning/data/streaming/data_processor.py

```diff
@@ -202,37 +203,48 @@ def _upload_fn(upload_queue: Queue, remove_queue: Queue, cache_dir: str, output_
             remove_queue.put([local_filepath])
 
 
-def _associated_items_to_workers(num_workers: int, user_items: List[Any]) -> Tuple[List[int], List[List[Any]]]:
+def _associated_items_to_workers(
+    num_workers: int, user_items: List[Any], weights: Optional[List[int]] = None
+) -> List[List[Any]]:
     # Associate the items to the workers based on number of nodes and node rank.
+    weights = [1] * len(user_items) if weights is None else weights
     num_nodes = _get_num_nodes()
-    current_node_rank = _get_node_rank()
-    node_size = len(user_items) // num_nodes
-    workers_user_items = []
-    begins = []
-    for node_rank in range(num_nodes):
-        if node_rank != current_node_rank:
+    node_rank = _get_node_rank()
+    world_size = num_nodes * num_workers
+
+    worker_items, worker_weights = _pack_greedily(items=user_items, weights=weights, num_bins=world_size)
+    worker_ids_this_node = range(node_rank * num_workers, (node_rank + 1) * num_workers)
+
+    for worker_id, size in worker_weights.items():
+        if worker_id not in worker_ids_this_node:
             continue
-        is_last_node = node_rank == num_nodes - 1
-        start_node = node_rank * node_size
-        end_node = len(user_items) if is_last_node else (node_rank + 1) * node_size
-        node_user_items = user_items[start_node:end_node]
-        worker_size = len(node_user_items) // num_workers
-        for worker_idx in range(num_workers):
-            is_last = worker_idx == num_workers - 1
-            begin = worker_idx * worker_size
-            end = len(node_user_items) if is_last else (worker_idx + 1) * worker_size
-            workers_user_items.append(node_user_items[begin:end])
-            begins.append(begin)
-        return begins, workers_user_items
-    raise RuntimeError(f"The current_node_rank {current_node_rank} doesn't exist in {num_nodes}.")
+        print(f"Worker {worker_id} gets {size / 1e6:.1f} MB ({len(worker_items[worker_id])} files)")
+
+    return [worker_items[worker_id] for worker_id in worker_ids_this_node]
+
+
+def _get_item_filesizes(items: List[Any], base_path: str = "") -> List[int]:
+    """Computes the total size in bytes of all file paths for every datastructure in the given list."""
+    item_sizes = []
+    for item in items:
+        flattened_item, spec = tree_flatten(item)
+
+        num_bytes = 0
+        for index, element in enumerate(flattened_item):
+            if isinstance(element, str) and element.startswith(base_path) and os.path.exists(element):
+                file_bytes = os.path.getsize(element)
+                if file_bytes == 0:
+                    raise RuntimeError(f"The file {element} has 0 bytes!")
```

Comment:
Should we raise or ignore here? Maybe just warn?

Reply:
Since this runs at the very beginning of processing, I think an error is appropriate. If a file in the dataset has 0 bytes, I think this should alarm the user to double check. Either they didn't assemble the dataset correctly, or if it can really be ignored, they should just delete the 0 bytes file. 

I'm only handling this edge case because in `_pack_greedily` I want all weights to be > 0. I could also relax it so allow weights to be 0, but it didn't appear realistic to me.

Reply:
The fact that it would fail immediately makes sense to me. All good then